packages/arb-validator/structures/node.go

```diff
@@ -243,27 +333,32 @@ func (node *Node) setHash(nodeDataHash common.Hash) {
 	node.hash = hash
 }
 
-func (node *Node) MarshalForCheckpoint(ctx *checkpointing.CheckpointContext) *NodeBuf {
+func Link(nd *Node, prev *Node) error {
+	if nd.prevHash != prev.hash {
+		return errors.New("node is not parent")
+	}
+	nd.prev = prev
+	prev.successorHashes[nd.linkType] = nd.hash
+	return nil
+}
+
+func (node *Node) MarshalForCheckpoint(ctx *ckptcontext.CheckpointContext, includeMachine bool) *NodeBuf {
 	var machineHash *common.HashBuf
-	if node.machine != nil {
+	if includeMachine && node.machine != nil {
```

Comment:
Might be good to add a comment explaining the decision to not checkpoint VM state for confirmed nodes.

Reply:
I'll add a comment to where the serialization is invoked rather than here, but I definitely agree there should be some explanation.